crates/fda/src/bench.rs

```diff
@@ -0,0 +1,495 @@
+use feldera_rest_api::Client;
+use log::{error, info, warn};
+use serde::{Deserialize, Serialize};
+use serde_json::{Map, Value};
+use std::cmp;
+use tabled::builder::Builder;
+use tabled::settings::Style;
+use tokio::time::{sleep, Duration, Instant};
+
+use crate::cli::{OutputFormat, PipelineAction};
+
+pub fn human_readable_bytes(n: i64) -> String {
+    const DELIMITER: f64 = 1024_f64;
+    const UNITS: &[&str] = &["B", "KiB", "MiB", "GiB", "TiB"];
+
+    let n: f64 = n as f64;
+    assert!(n >= 0f64, "No negative bytes");
+
+    let n = n.abs();
+    if n < 1_f64 {
+        return format!("{} {}", n, "B");
+    }
+    let exp = cmp::min(
+        (n.ln() / DELIMITER.ln()).floor() as i32,
+        (UNITS.len() - 1) as i32,
+    );
+    let bytes = format!("{:.2}", n / DELIMITER.powi(exp))
+        .parse::<f64>()
+        .unwrap()
+        * 1_f64;
+    let unit = UNITS[exp as usize];
+    format!("{} {}", bytes, unit)
+}
+
+/// Raw metrics collected from a single measurement
+#[derive(Debug)]
+struct RawMetrics {
+    rss_bytes: i64,
+    uptime_msecs: i64,
+    incarnation_uuid: String,
+    storage_bytes: i64,
+    total_processed_records: i64,
+    input_bytes: i64,
+    input_errors: bool,
+}
+
+/// JSON representation of input connector metrics
+#[derive(Debug, Deserialize)]
+#[allow(dead_code)]
+struct InputMetrics {
+    buffered_records: i64,
+    end_of_input: bool,
+    num_parse_errors: i64,
+    num_transport_errors: i64,
+    total_bytes: i64,
+    total_records: i64,
+}
+
+/// JSON representation of an input connector
+#[derive(Debug, Deserialize)]
+#[allow(dead_code)]
+struct InputConnector {
+    barrier: bool,
+    config: Map<String, Value>,
+    endpoint_name: String,
+    fatal_error: Option<String>,
+    metrics: InputMetrics,
+    paused: bool,
+}
+
+/// JSON representation of pipeline stats
+#[derive(Debug, Deserialize)]
+#[allow(dead_code)]
+struct PipelineStats {
+    global_metrics: Map<String, Value>,
+    inputs: Vec<InputConnector>,
+    outputs: Vec<Value>,
+    suspend_error: Option<Value>,
+}
+
+impl From<&Map<String, Value>> for RawMetrics {
+    fn from(stats: &Map<String, Value>) -> Self {
+        let pipeline_stats: PipelineStats =
+            serde_json::from_value(Value::Object(stats.clone())).unwrap();
+        let global_metrics = &pipeline_stats.global_metrics;
+
+        // Check for input errors
+        let input_errors = pipeline_stats.inputs.iter().any(|input| {
+            input.fatal_error.is_some()
+                || input.metrics.num_parse_errors > 0
+                || input.metrics.num_transport_errors > 0
+        });
+
+        // Calculate total input bytes
+        let input_bytes = pipeline_stats
+            .inputs
+            .iter()
+            .map(|input| input.metrics.total_bytes)
+            .sum();
+
+        Self {
+            rss_bytes: global_metrics
+                .get("rss_bytes")
+                .and_then(|v| v.as_i64())
```

Comment:
if you unwrap before would this be a map?

Reply:
?